### DISCUSSION TITLE
aes(color = after_stat(change)) error for even when color is *set*

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#37] [Q&A]
This behavior seems surprising to me: that layer w/o no variable 'change' produced in the stat stage errors even when the aes is *set*.  Feels like it shouldn't be the case.   

error produced at last_plot() +  aes(color = sign(after_stat(change))).   We can make it go away by wiping last layer, stat_summary(color = "goldenrod"), and adding it again, but with mapping = aes(color = NULL) declared locally.  Thanks!

``` r
library(tidyverse)
compute_panel_pairs <- function(data, scales){
  
  data %>% 
    mutate(group_label = group) %>% 
    group_by(group) %>% 
    arrange(x) %>% 
    mutate(change = y - lag(y)  ) %>% 
    fill(change, .direction = "up")  
  
}

StatPairs <- ggproto("StatPairs", Stat,
                     compute_panel = compute_panel_pairs,
                     default_aes = aes(color = after_stat(change)))

gapminder::gapminder %>% 
  filter(year %in% c(1992, 2002)) %>% 
  ggplot() + 
  aes(x = year, 
      y = gdpPercap, 
      group = country) + 
  geom_point(stat = StatPairs) +
  geom_path(stat = StatPairs)
```

![](https://i.imgur.com/4LMOJmf.png)

``` r


last_plot()  +
  geom_text(
            stat = StatPairs,
            aes(label = country), 
            hjust = "outward", 
            check_overlap = T
            )
```

![](https://i.imgur.com/UkcEpcJ.png)

``` r

last_plot() + 
  aes(x = as.factor(year))
```

![](https://i.imgur.com/byrl9AJ.png)

``` r

last_plot() + 
  stat_summary(geom = "point", 
               color = "goldenrod", 
               fun = "mean",
               aes(group = NULL, 
                   # color = NULL  # why do I have to do this? When I have color *set* above?
                   ), 
               size = 4)
```

![](https://i.imgur.com/BmRSBom.png)

``` r


last_plot() + 
  aes(color = sign(after_stat(change))) 
#> Error in `stat_summary()`:
#> ! Problem while mapping stat to aesthetics.
#> ℹ Error occurred in the 4th layer.
#> Caused by error:
#> ! object 'change' not found
```

``` r

ggwipe::last_plot_wipe_last() + 
  stat_summary(geom = "point", 
               color = "goldenrod", 
               fun = "mean",
               aes(group = NULL, 
                   color = NULL  # why do I have to do this? When I have color *set* above?
                   ), 
               size = 4)
```

![](https://i.imgur.com/bBfYHCB.png)

<sup>Created on 2024-09-05 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by teunbrand** [comment]
The following line is key:

> #> Error in `stat_summary()`:

The summary stat does not produce a `change` column, so trying to set `+ aes(color = after_stat(sign(change)))` globally will fail for that layer.

> `# why do I have to do this? When I have color *set* above?`

This is also for the reason above. The local `aes(colour = NULL)` takes precedence over the global `aes(colour = sign(after_stat(change)))`.

**Comment by EvaMaeRey** [comment]
Thanks for the discussion. it just feels funny (I know there might not be anything worth doing about it) because of the unmapped aesthetic declaration, color = "goldenrod" and the fact that that local declaration usually ultimately governs color for the layer.  Like if we did global aes(color = country), that country info doesn't end up in the visual display.  

So it doesn't seem like the layer should 'care' about the global aes declaration when there's an unmapped, local aes declaration in play.  With the syntax that does work for stat_summary(), it's like I'm saying 'don't map color' twice.  I know @yjunechoe talks about using after_stat() globally, and just advises against.  But I'm addicted to global mapping.  😔

**Reply by teunbrand** [reply]
I rcently found out that in the layer `geom_point(aes(colour = foo), colour = "red")`, the `colour = foo` is actually evaluated and kept for a big part of plot building. It is only in `Layer$compute_geom_2()` that `colour = "red"` overrides the mapping. So local fixed aesthetics don't prevent homonymous mappings from occurring.

Also not using `after_stat()` in global aesthetics is good advice if your plot >1 layer.

**Reply by EvaMaeRey** [reply]
Ah, so carried along for a long time...  So interesting! I think for this specific case, since after_stat is being used in three layers (line, point, text), and I might want to map that globally, then doing a local aes(color = I("goldenrod")) for stat_summary might be the way, to avoid saying things twice.   And probably right general advice is June's - don't go global w after_stat().

  stat_summary(geom = "point", 
               fun = "mean",
               aes(group = NULL, 
                      color = I("goldenrod")), 
               size = 4)

Just double checked...  works.

**Reply by teunbrand** [reply]
Oh yeah, smart! I forgot this is exactly the type of situation where `I()` can come to the rescue 😅

**Reply by yjunechoe** [reply]
This is a bit of a tangent but on the note of "hard-coding aes" and "homonymous mappings", I always found this behavior of `stat_summary(fun.data)` interesting:

Base plot:

```
library(ggplot2)

p1 <- ggplot(mtcars, aes(cyl, mpg)) +
  stat_summary(
    aes(fill = factor(am)),
    position = "dodge",
    geom = "bar",
    fun.data = ~ mean_se(.x)
  )
p1
```

![image](https://github.com/user-attachments/assets/8df7d42d-c0d3-4457-beca-623b6cf2cf00)


Change up `fun.data` to hard-code fill to pink:

```r
p2 <- ggplot(mtcars, aes(cyl, mpg)) +
  stat_summary(
    aes(fill = factor(am)),
    position = "dodge",
    geom = "bar",
    fun.data = ~ mean_se(.x) |> transform(fill = "pink") # hard-code color via `fun.data`
  )
p2
```

![image](https://github.com/user-attachments/assets/e6bf34ee-5ed9-43c3-a61d-b215e229802d)

The attempt to hard-code via fun.data is catastrophic because it bypasses the existing mechanism for dealing with overlapping aes specifications, and gets handled by `merge()` instead. Consequently, the plot doesn't resolve overlapping `fill` aes values; instead it produces "repaired" column names for the overlapping columns (`fill.x` and `fill.y` - one from aes mapping and one from `fun.data`). To top it all off, when the plot comes back to the layer data later, it notices that `fill` isn't present, so it adds a `fill` column with the default values. All of that giving you 3 copies of the fill column:

```r
layer_data(p2, 1) |> 
  dplyr::select(starts_with("fill"))
#>   fill.x fill.y      fill
#> 1   pink      0 #595959FF
#> 2   pink      0 #595959FF
#> 3   pink      0 #595959FF
#> 4   pink      1 #595959FF
#> 5   pink      1 #595959FF
#> 6   pink      1 #595959FF
```

**Reply by teunbrand** [reply]
Huh would you look at that! This seems like a bona fide bug to me. What should happen is that `fill = "pink"` should become mapped to a scale and all bars should be displayed with whatever the first colour in `scale_fill_discrete()` is.

**Reply by EvaMaeRey** [reply]
> Oh yeah, smart! I forgot this is exactly the type of situation where `I()` can come to the rescue 😅

`I()` is great...  Pretty sure I started using after last ggplot2 release notes!

**Reply by yjunechoe** [reply]
@teunbrand I'd be satisfied if at least the new summary stat did away with this behavior 😁. Probably a bug for `stat_summary()` too but I enjoy the vibe of `fun.data` being a kind of a backdoor danger zone 😆

**Reply by teunbrand** [reply]
I like that `stat_summary()` sees this conflict and decides to placate neither party and confidently proceeds to do its own thing incorrectly 

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExZXRpNWl1MWV1dW5ma3RscGpudG5nd3BzdnNkeDhoZXNkZzNscWdnMCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/1eKbAKgocJCta/giphy.gif)

**Reply by yjunechoe** [reply]
Yeah it's the "hmm I see you don't have a `fill` for this layer lemme help you with that" attitude that gets me 😂

**Reply by yjunechoe** [reply]
Also!!! I was wondering why this sounded so familiar when I was reading it:

> I rcently found out that in the layer geom_point(aes(colour = foo), colour = "red"), the colour = foo is actually evaluated and kept for a big part of plot building. It is only in Layer$compute_geom_2() that colour = "red" overrides the mapping. So local fixed aesthetics don't prevent homonymous mappings from occurring.

Turns out that this was actually one of the first internals quirk I investigated with ggtrace, when I had _just_ started to work on it back in 2021. I think I presented these slides for @EvaMaeRey and Peter Higgins - one of the precursors to the current ggplot extenders meet-ups!

https://yjunechoe.github.io/ggtrace-talk/

![image](https://github.com/user-attachments/assets/373748bd-6068-4204-ad01-6125d9f7bdf3)

**Reply by teunbrand** [reply]
Nice, an golden oldie for you then! I found this out while doing the debug for https://github.com/jamesotto852/ggdensity/pull/35. I should've known this :grin:

**Reply by EvaMaeRey** [reply]
Almost three years ago! 


In addition to ggtrace, I think it would be really instructive to step through 'tiered succession of calls' that gets data to render-ready.  ![image](https://github.com/user-attachments/assets/4b263b0f-dbb7-4a87-ad6b-ab9feb175c54)

https://github.com/EvaMaeRey/mytidytuesday/blob/master/2023-09-15-ggplot2-stat-cascades/ggplot2_stat_cascades.Rmd  

I tried this a while a go ...

library(ggplot2)
mtcars |> 
  dplyr::rename(y = mpg, x = vs) |> 
  StatCount$setup_data() |> 
  StatCount$setup_params() |> # complains
  StatCount$compute_layer() |> # complains
  StatCount$compute_panel() |> # complains
  StatCount$compute_group() 


... and fail even in the Stat space, which is my favorite, but haven't looked at in a while so maybe I could make it work if taking another look.

**Reply by teunbrand** [reply]
Gina, I don't mean to be rude, but I have no idea why you expect that code to work :p I think what Thomas is getting at with the tiered succession of calls is that `compute_layer()` calls `compute_panel()` calls `compute_group()`.

**Reply by yjunechoe** [reply]
I initially had this confusion too, though I forget what exactly got me trapped in that tunnel vision. It took me a while to get the precise intent behind _tiered succession_ as:

```
compute_layer()
└──compute_panel()
    └── compute_group()
```

For what it's worth, I still think the "internals as a data pipeline of `data %>% f1() %>% f2() %>% ...`" is useful as a high-level mental model 😄 . It's just that in the narrowest sense of that idea, this only really happens at one very specific (and kinda useless, in a practical sense) level of abstraction.

```r
library(ggtrace)
library(ggplot2)

build <- as.list(body(ggplot2:::ggplot_build.ggplot))
layer_methods <- sapply(build, deparse1) |>
  grep(x = _, "l\\$\\w+", value = TRUE) |>
  gsub(x = _, ".*l(\\$\\w+).*", "ggplot2:::Layer\\1")
names(layer_methods) <- gsub(x = layer_methods, ".*\\$", "")

p <- ggplot(mtcars, aes(cyl)) +
  stat_count()

stat_count_Layer <- lapply(
  layer_methods,
  \(x) {
    rlang::inject(
      capture_fn(p, !!rlang::parse_expr(x), cond = 1)
    )
  }
)
p_Layout <- list(setup = capture_fn(p, Layout$setup))

layer_data(p, 1)
#>    y count    prop x flipped_aes PANEL group ymin ymax xmin xmax colour
#> 1 11    11 0.34375 4       FALSE     1    -1    0   11  3.1  4.9     NA
#> 2  7     7 0.21875 6       FALSE     1    -1    0    7  5.1  6.9     NA
#> 3 14    14 0.43750 8       FALSE     1    -1    0   14  7.1  8.9     NA
#>        fill linewidth linetype alpha
#> 1 #595959FF       0.5        1    NA
#> 2 #595959FF       0.5        1    NA
#> 3 #595959FF       0.5        1    NA
```

```r
mtcars |>
  stat_count_Layer$layer_data(plot_data = _) |>
  stat_count_Layer$setup_layer(data = _) |>
    # Detour around Layout$setup()
    list() |>
    p_Layout$setup(data = _) |>
    el(1) |>
  stat_count_Layer$compute_aesthetics(data = _) |>
  stat_count_Layer$compute_statistic(data = _) |>
  stat_count_Layer$map_statistic(data = _) |>
  stat_count_Layer$compute_geom_1(data = _) |>
  stat_count_Layer$compute_geom_2(data = _) |>
  stat_count_Layer$finish_statistics(data = _)
#>    y count    prop x flipped_aes PANEL group ymin ymax xmin xmax colour
#> 1 11    11 0.34375 4       FALSE     1    -1    0   11  3.1  4.9     NA
#> 2  7     7 0.21875 6       FALSE     1    -1    0    7  5.1  6.9     NA
#> 3 14    14 0.43750 8       FALSE     1    -1    0   14  7.1  8.9     NA
#>        fill linewidth linetype alpha
#> 1 #595959FF       0.5        1    NA
#> 2 #595959FF       0.5        1    NA
#> 3 #595959FF       0.5        1    NA
```

I almost wanted to show exactly this pipeline code above in my rstudioconf talk but ultimately decided against it as it was too complicated to set up and too confusing to explain - so I went with the simplier pseudocode below instead. Fully admit that this is a "leaky" abstraction, and perhaps only useful as a training wheel - kinda like how in science you get introduced to the intuitive and familiar model of newtonian physics _first_, and _then_ you're supposed to unlearn that when you get to quantum mechanics (or so I hear - not a physicist 😆).

![image](https://github.com/user-attachments/assets/d6ba4961-ef16-4d42-91bc-cc9d6f89141c)

**Reply by EvaMaeRey** [reply]
@teunbrand I'm going to take it as a point of pride that I'm bringing mind boggling ideas here (if off base)!    @yjunechoe Thank you so much for sharing this. Very exciting! I love the stepping through that you do. 

I think it would be fun/instructive to walk- and talk- through at some point.  Like have someone a bit more internals-naive (can be me) take a shot at narration, but be open to correction along the way.  
 https://evamaerey.github.io/mytidytuesday/2024-09-09-junes-gg-cascade/cascade_chunk_flipbook.html

**Reply by yjunechoe** [reply]
Cool to see this turned into a flipbook! I will say though that this is *still* with a layer of abstraction on top, so I'm not confident about its usefulness outside of my own personal understanding, but I do like the idea of walking through the steps of ggplot build.

Along those lines, I think it may be more faithful and fruitful to just `n` over the lines of `ggplot2:::ggplot_build.ggplot` in the interactive debugger and narrate each step there, since that preserves *exactly* how things play out:

```r
debugonce(ggplot2:::ggplot_build.ggplot)
plot(p)
```

**Reply by EvaMaeRey** [reply]
Sounds like a good plan.  Thanks for the suggestion!
